src/pcre2_compile.c

```diff
@@ -1900,17 +1903,20 @@ else
 
       /* As we know we are at a digit, the only possible error from
       read_number() is a number that is too large to be a group number. In this
-      case we fall through handle this as not a group reference. If we have
-      read a small enough number, check for a back reference.
+      case we treat the group number as too-large (since it may be larger than
+      INT_MAX we cannot return it for the caller to check).
 
       \1 to \9 are always back references. \8x and \9x are too; \1x to \7x
       are octal escapes if there are not that many previous captures. */
 
-      if (read_number(&ptr, ptrend, -1, INT_MAX/10 - 1, 0, &s, errorcodeptr) &&
-          (s < 10 || oldptr[-1] >= CHAR_8 || s <= (int)cb->bracount))
+      if (!read_number(&ptr, ptrend, -1, MAX_GROUP_NUMBER, 0, &s, errorcodeptr))
```

Comment:
I took the opportunity to reduce the limit from INT_MAX/10 to MAX_GROUP_NUMBER, although it makes no difference really.

The only (minor) impact is that the error is now reported on the location of the first character exceeding MAX_GROUP_NUMBER, which makes the location of the error more consistent with the other places `read_number()` is used. But arguably anyway, `read_number()` should perhaps advance the pointer to end of the run of digits when reporting a "too many digits" error, instead of just leaving pointer in the middle of the digit sequence.

Reply:
What is the problem with ALT_BSUX here? That option is supposed to apply only to \U, \u and \x.

Reply:
After looking at it further, I think the limit should be reverted to INT_MAX/10 - 1.

We would get into this code also if the octal we are reading is next to other numbers and will incorrectly fail (ex: /\\214748364/), also the check after for the returned value being larger than MAX_GROUP_NUMBER makes no sense otherwise.

Reply:
> After looking at it further, I think the limit should be reverted to INT_MAX/10 - 1.

I'm not convinced. INT_MAX is a completely arbitrary non-semantic limit, that's just a random number unrelated to the true limit of MAX_GROUP_NUMBER. Exposing the size of `int` (so that 5,000,000,000 is treated differently to 1,000,000,000) is strange - the regex grammar doesn't in any way depend on the size of int!

The original code was extremely close to being correct: it used `read_number` to peek for a decimal, without touching the `errorcodeptr`. We don't need to chage that.

Reply:
> The original code was extremely close to being correct: it used read_number to peek for a decimal, without touching the errorcodeptr. We don't need to chage that.

The bug was likely introduced in 8144ae04, without realizing that because `errorcodeptr` was no longer being set, the break when `c == CHAR_8` was no longer propagating that upstream.